### PR TITLE
feat: migrate node operations to MVI

### LINKED ISSUE #97: Feature: MVI Editor Operations
Implement the MVI Editor Operations to update node type, text, and modifiers safely via intents. Ensure children are preserved when type is updated to a compatible container.

### PR DESCRIPTION
**Pull request by jesusdmedinac**
Resolves #97. Migrated node operations (Update NodeType, Update Text, Modifiers CRUD) to MVI using EditorIntent and EditorState. Ensures immutability via proper node copying and tree recreation. Covered with TDD.

## Summary by CodeRabbit

* **New Features**
  * Added editor support for changing node types, updating text, and managing modifiers.
  * Node type selection now displays only compatible options.
  * Compatible child content is preserved when changing node types.
  * Added default values for newly created modifier operations.
  * Recursive updates now preserve unaffected parts of the editor tree.

* **Documentation**
  * Updated MVI editor operation documentation and completion progress.

* **Tests**
  * Added coverage for node updates, modifier CRUD, compatibility rules, edge cases, and structural preservation.

### CODE REVIEW

File: composy/src/commonMain/kotlin/com/jesusdmedinac/compose/sdui/presentation/screenmodel/ComposeNodeUtils.kt [hunk 1]
```diff
@@ -50,3 +50,31 @@ fun ComposeNode.nodeExists(targetId: String): Boolean {
     if (this.id == targetId) return true
     return this.children().any { it.nodeExists(targetId) }
 }
+
+/**
+ * Traverses the tree and applies [transform] to the node with [targetId].
+ */
+fun ComposeNode.updateNodeRecursive(targetId: String, transform: (ComposeNode) -> ComposeNode): ComposeNode {
+    if (this.id == targetId) return transform(this)
```
[the remark this answers is not in the document]
Comment (jesusdmedinac) [reply composy/src/commonMain/kotlin/com/jesusdmedinac/compose/sdui/presentation/screenmodel/ComposeNodeUtils.kt:76]: Fixed in 06fcfce. Went with structural sharing instead of the `nodeExists` pre-check: the recursion now returns `this` unchanged when no descendant changed, so untouched subtrees keep instance identity with a single traversal (the pre-check would re-walk each subtree). Covered by `testUpdateNodeRecursive_untouchedSiblingsKeepIdentity`.

File: composy/src/commonMain/kotlin/com/jesusdmedinac/compose/sdui/presentation/screenmodel/EditorScreenModel.kt [hunk 2]
```diff
@@ -64,23 +64,68 @@ class EditorScreenModel : ScreenModel, EditorIntent.Visitor {
...
+            val updatedRoot = state.rootNode.updateNodeRecursive(intent.id) { node ->
+                val newOperations = node.composeModifier.operations.toMutableList()
+                if (intent.index in newOperations.indices) {
+                    newOperations[intent.index] = intent.operation
+                }
+                node.copy(composeModifier = com.jesusdmedinac.jsontocompose.model.ComposeModifier(newOperations))
+            }
+            state.copy(rootNode = updatedRoot)
+        }
     }
```
[the remark this answers is not in the document]
Comment (jesusdmedinac) [reply composy/src/commonMain/kotlin/com/jesusdmedinac/compose/sdui/presentation/screenmodel/EditorScreenModel.kt:117]: Fixed as suggested in 06fcfce, with a regression test (`testUpdateModifier_outOfBoundsIndex_keepsRootNodeInstance`) asserting the root node keeps instance identity on an out-of-bounds index.

File: composy/src/commonMain/kotlin/com/jesusdmedinac/compose/sdui/presentation/screenmodel/EditorScreenModel.kt [hunk 3]
```diff
@@ -64,23 +64,68 @@ class EditorScreenModel : ScreenModel, EditorIntent.Visitor {
...
+            val updatedRoot = state.rootNode.updateNodeRecursive(intent.id) { node ->
+                val newOperations = node.composeModifier.operations.toMutableList()
+                if (intent.index in newOperations.indices) {
+                    newOperations.removeAt(intent.index)
+                }
+                node.copy(composeModifier = com.jesusdmedinac.jsontocompose.model.ComposeModifier(newOperations))
+            }
+            state.copy(rootNode = updatedRoot)
+        }
     }
```
[the remark this answers is not in the document]
Comment (jesusdmedinac) [reply composy/src/commonMain/kotlin/com/jesusdmedinac/compose/sdui/presentation/screenmodel/EditorScreenModel.kt:134]: Fixed as suggested in 06fcfce, with a regression test (`testDeleteModifier_outOfBoundsIndex_keepsRootNodeInstance`).

File: composy/src/commonTest/kotlin/com/jesusdmedinac/compose/sdui/presentation/screenmodel/EditorScreenModelTest.kt [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+import com.jesusdmedinac.jsontocompose.model.NodeProperties
+import com.jesusdmedinac.jsontocompose.model.ComposeModifier
+import com.jesusdmedinac.jsontocompose.modifier.ModifierOperation
+import kotlin.test.Test
+import kotlin.test.assertEquals
+import kotlin.test.assertIs
+import kotlin.test.assertTrue
+import kotlin.test.assertNotNull
+
+class EditorScreenModelTest {
```
[the remark this answers is not in the document]
Comment (jesusdmedinac) [reply composy/src/commonTest/kotlin/com/jesusdmedinac/compose/sdui/presentation/screenmodel/EditorScreenModelTest.kt:15]: The Gradle daemon failure happened in CodeRabbit's own sandbox, not in this repo's CI — the "Build and Deploy to Cloudflare" check passed on this PR, and `./gradlew :composy:desktopTest` (which compiles `:composy`) passes locally as of 06fcfce. No code change needed.

File: PROGRESS.md [hunk 5]
```diff
@@ -14,14 +14,14 @@
 | 2   | [phase-actions-system](#2-phase-actions-system)                   | `library/`    | 37        | 37        | **100%** |
 | 3   | [phase-demo-showcase](#3-phase-demo-showcase)                     | `composeApp/` | 36        | 36        | **100%** |
 | 4   | [phase-3-expand-library](#phase-3-expand-library)               | `library/`    | 239       | 239       | **100%** |
-| 5   | [phase-1.5-mvi-architecture](#phase-15-mvi-architecture)        | `composy/`    | 12        | 6         | **50%**  |
+| 5   | [phase-1.5-mvi-architecture](#phase-15-mvi-architecture)        | `composy/`    | 12        | 9         | **75%**  |
 | 6   | [phase-2-editor-mvp](#phase-2-editor-mvp)                       | `composy/`    | 58        | 19        | **32%**  |
 | 7   | [phase-4-semantics-testability](#phase-4-semantics-testability) | `library/`    | 30        | 0         | **0%**   |
 | 8   | [phase-3-differentiators](#phase-3-differentiators)             | Multi-module  | 37        | 0         | **0%**   |
```
[the remark this answers is not in the document]
Comment (jesusdmedinac) [reply PROGRESS.md:20]: Fixed in 06fcfce. The root cause was bigger than the lint warning: a corrupted duplicate "Recommended Development Order" block after the progress table, plus table links pointing at completed-phase sections that were removed from the file. Deleted the orphan block and de-linked those three phases; the remaining anchors resolve and markdownlint no longer reports MD051 on PROGRESS.md.

File: library/src/commonMain/kotlin/com/jesusdmedinac/jsontocompose/model/ComposeType.kt [hunk 6]
```diff
@@ -195,4 +195,16 @@ enum class ComposeType {
...
         else -> false
     }
+
+    /**
+     * Returns a list of ComposeTypes that this type can safely be converted to.
+     * Layouts can be converted to other Layouts.
+     * Single child containers can be converted to other single child containers.
+     * Otherwise, returns all types.
+     */
+    fun compatibleTypes(): List<ComposeType> = when {
```
Comment (jesusdmedinac) [inline library/src/commonMain/kotlin/com/jesusdmedinac/jsontocompose/model/ComposeType.kt]: **Medium — multi-slot containers fall through to `else` and can silently drop children.**

`Scaffold`, `AlertDialog`, `TopAppBar`, `ListItem`, and `SearchBar` all hold child nodes but return `false` from both `isLayout()` and `hasChild()`, so they land in `else -> ComposeType.entries`. That means a `Scaffold` with a content child, top bar, and FAB can be converted to `Text` — and every one of those children is silently discarded, which contradicts the scenario's "children should be preserved in the new container if compatible".

The same asymmetry exists in the other direction: an *empty* `Column` can't become `Text` (blocked by `isLayout()`-only filtering), while a fully populated `Scaffold` can become anything.

Suggestion: decide compatibility from whether the node *currently has children* rather than (or in addition to) the static type category — e.g. a node with children may only convert to types that can hold them, and a childless node may convert to anything. At minimum, add the multi-slot container types to the guarded set.

Comment (jesusdmedinac) [inline library/src/commonMain/kotlin/com/jesusdmedinac/jsontocompose/model/ComposeType.kt]: **Note — no UI consumes this yet, so the third scenario clause is unimplemented.**

The feature file (updated in this PR) says "incompatible types should not be presented in the UI", but `compatibleTypes()` is only used by the ScreenModel guard, and no UI code in `composy` dispatches any `EditorIntent` yet (`onIntent` calls exist only inside the `screenmodel` package). The remaining "MVI UI State" scenarios cover panel toggles/viewport/search, not the type-selector wiring — so this clause isn't tracked anywhere once this PR merges. Suggest either wiring the type dropdown to filter by `compatibleTypes()` here, or splitting that clause into its own scenario/issue so it doesn't get lost.

Related placement nit: `compatibleTypes()` and `ModifierOperation.createDefaultOperation()` are editor-authoring concerns added to the `library` module's public API, while their sibling `createDefaultProperties()` lives in `composy`. Consider keeping all three on the same side of the boundary.

File: composy/src/commonMain/kotlin/com/jesusdmedinac/compose/sdui/presentation/screenmodel/EditorScreenModel.kt [hunk 7]
```diff
@@ -64,23 +64,76 @@ class EditorScreenModel : ScreenModel, EditorIntent.Visitor {
...
 
     override fun visit(intent: EditorIntent.UpdateNodeText) {
-        // To be implemented in next features
+        reduce { state ->
+            val updatedRoot = state.rootNode.updateNodeRecursive(intent.id) { node ->
+                val newProps = when (val props = node.properties) {
+                    is com.jesusdmedinac.jsontocompose.model.NodeProperties.TextProps -> props.copy(text = intent.text)
+                    else -> props
```
Comment (jesusdmedinac) [inline composy/src/commonMain/kotlin/com/jesusdmedinac/compose/sdui/presentation/screenmodel/EditorScreenModel.kt]: **Minor — non-Text nodes get a semantic no-op that still rebuilds the path.**

When `props` isn't `TextProps`, this returns the same properties instance but still calls `node.copy(...)`, producing a new node and forcing `updateNodeRecursive` to reallocate the root-to-target path (and emit a new state) for a no-op. Returning `node` directly in the `else` case would keep instance identity, matching the pattern the `UpdateModifier`/`DeleteModifier` handlers already use for their out-of-bounds case:

```kotlin
val props = node.properties
if (props !is NodeProperties.TextProps) return@updateNodeRecursive node
node.copy(properties = props.copy(text = intent.text))
```

File: composy/src/commonMain/kotlin/com/jesusdmedinac/compose/sdui/presentation/screenmodel/ComposeTypeExtensions.kt [hunk 8]
```diff
@@ -71,30 +71,35 @@ import com.jesusdmedinac.jsontocompose.model.ComposeType.TimePicker
...
  * Creates a default, non-null properties instance for this component type.
+ * Optionally preserves existing children when changing a node's type.
  */
-fun ComposeType.createDefaultProperties(): NodeProperties = when (this) {
-    Column, LazyColumn -> NodeProperties.ColumnProps()
-    Row, LazyRow -> NodeProperties.RowProps()
-    Box -> NodeProperties.BoxProps()
+fun ComposeType.createDefaultProperties(
+    preservedChildren: List<ComposeNode>? = null,
+    preservedChild: ComposeNode? = null
```
Comment (jesusdmedinac) [inline composy/src/commonMain/kotlin/com/jesusdmedinac/compose/sdui/presentation/screenmodel/ComposeTypeExtensions.kt]: **Nit — `preservedChild` is redundant with `preservedChildren`.**

The only caller (`UpdateNodeType`) passes `preservedChild = node.children().firstOrNull()` alongside `preservedChildren = node.children().takeIf { it.isNotEmpty() }`, so every `preservedChild ?: preservedChildren?.firstOrNull()` fallback below is dead — the two parameters always agree. Dropping `preservedChild` and using `preservedChildren?.firstOrNull()` in the single-child branches would simplify the signature without changing behavior.

**Review by jesusdmedinac — COMMENTED** [review]
Reviewed against the PR's purpose (#97: migrate UpdateNodeType, UpdateNodeText, and Modifiers CRUD to MVI). The migration itself is solid: all five handlers follow the same `reduce` + `updateNodeRecursive` pattern, immutability is preserved with structural sharing, and the three scenarios have test coverage. Three findings inline:

1. **Medium** — `compatibleTypes()` silently drops children for multi-slot containers (Scaffold, AlertDialog, TopAppBar, ListItem, SearchBar) because they fall into the `else -> ComposeType.entries` branch.
2. **Note on scope** — the scenario clause "incompatible types should not be presented in the UI" has no implementation yet: nothing in `composy` UI dispatches any `EditorIntent` or consumes `compatibleTypes()`. Fine if UI wiring is deferred, but it isn't covered by the remaining "MVI UI State" scenarios either, so it should be tracked explicitly.
3. **Minor** — `UpdateNodeText` on a non-Text node rebuilds the root-to-target path for a semantic no-op.

Also worth a follow-up test (not blocking): the incompatibility guard in `UpdateNodeType` (e.g. a `Column` with children dispatched to `Button` leaves the tree unchanged) is currently untested — `testUpdateNodeType_preservesChildrenIfCompatible` only covers the happy path.

File: library/src/commonMain/kotlin/com/jesusdmedinac/jsontocompose/model/ComposeType.kt (hunk 6 above)
Comment (jesusdmedinac) [reply library/src/commonMain/kotlin/com/jesusdmedinac/jsontocompose/model/ComposeType.kt]: Fixed in c79b7ac. Replaced the type-category check with a node-aware `ComposeNode.compatibleTypes()` (in `composy`'s `ComposeNodeUtils.kt`): a node with multiple children can only convert to multi-child layouts, one child also allows single-child containers, and childless nodes can convert to anything. Populated Scaffold/ListItem/etc. can no longer drop children, and the empty-Column asymmetry is gone. Known trade-off: a node with exactly one child can't convert *into* multi-slot types like Scaffold (neither `isLayout` nor `hasChild`) — conservative by design, no data loss. Covered by `ComposeNodeUtilsTest` plus guard tests in `EditorScreenModelTest`.

Comment (jesusdmedinac) [reply library/src/commonMain/kotlin/com/jesusdmedinac/jsontocompose/model/ComposeType.kt]: Addressed in c79b7ac: the type dropdown in `ComposeNodeEditor` now filters by `editingComposeNode.compatibleTypes()`, so incompatible types are no longer presented — this also protects the legacy `EditNodeScreenModel.onComposeTypeSelected` path the dropdown still uses (minor overlap with PR #87 to reconcile at merge). The broader MVI UI wiring gap is now tracked in #99. Placement nit also fixed: `compatibleTypes()` moved from `library` to `composy`, next to `createDefaultProperties`.

File: composy/src/commonMain/kotlin/com/jesusdmedinac/compose/sdui/presentation/screenmodel/EditorScreenModel.kt (hunk 7 above)
Comment (jesusdmedinac) [reply composy/src/commonMain/kotlin/com/jesusdmedinac/compose/sdui/presentation/screenmodel/EditorScreenModel.kt]: Fixed in c79b7ac — non-Text nodes now return `node` unchanged, matching the modifier handlers' no-op pattern. Regression test: `testUpdateNodeText_nonTextNode_keepsRootNodeInstance`.

File: composy/src/commonMain/kotlin/com/jesusdmedinac/compose/sdui/presentation/screenmodel/ComposeTypeExtensions.kt (hunk 8 above)
Comment (jesusdmedinac) [reply composy/src/commonMain/kotlin/com/jesusdmedinac/compose/sdui/presentation/screenmodel/ComposeTypeExtensions.kt]: Fixed in c79b7ac — dropped `preservedChild`; single-child branches use `preservedChildren?.firstOrNull()`.
